keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,838 @@
+it is possible to create new objects with invalid IP values (since it
+won't be possible any more).
+
+##### e2e tests
+
+No new tests, and we will remove `test/e2e/network/funny_ips.go`,
+which tests the behavior of kube-proxy when it sees
+Service/EndpointSlice objects with invalid IPs (since it won't be
+possible to create such objects any more). We already added [equivalent
+unit tests] to the iptables and nftables backends of kube-proxy (which
+actually caught a bug in kube-proxy that the e2e test hadn't).
+
+This means that we will lose testing for out-of-tree service proxy
+implementations (and out-of-tree NetworkPolicy implementations, etc),
+but there doesn't seem to be anything we can do about this: we can
+only have an e2e test of "what does the implementation do with bad IP
+values?" if we can create Pods / Services / EndpointSlices /
+NetworkPolicies / etc with bad IP values. But the whole point of this
+KEP is to make that impossible.
+
+(I guess possibly we might be able to set up the e2e test to be able
+to write the object directly into etcd, bypassing the apiserver? Or
+maybe add an "e2e test run" mode to the apiserver, that makes it
+possible to request that it skip validation for a particular
+Create/Update? Alternatively, this is an argument for eventually
+forcibly fixing all remaining invalid values in etcd, since it will no
+longer be possible to test behavior regarding them.)
+
+[equivalent unit tests]: https://github.com/kubernetes/kubernetes/pull/126203
+
+### Graduation Criteria
+
+#### Alpha
+
+- API warnings added.
+
+- Mandatory strict validation for `NetworkDeviceData`.
+
+- All kube-internal controllers only write out canonical IPs/CIDRs.
+
+- Updated validation available behind a feature gate, but disabled by
+  default.
+
+- Unit tests added/updated.
+
+#### Beta
+
+- Updated validation available behind a feature gate, enabled by
+  default.
+
+- Unit tests added/updated, old e2e and integration tests removed.
+
+#### GA
+
+- Allow time for feedback, etc
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A; No interactions between kubernetes components should be affected
+by the change, so skew doesn't matter.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: StrictIPCIDRValidation
+  - Components depending on the feature gate:
+    - kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+Yes, it changes the validation of multiple fields.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Nothing surprising.
+
+###### Are there any tests for feature enablement/disablement?
+
+No
```

Comment:
This is required for every feature. See my other comment about integration tests, that should be sufficient to ensure the previous/broken behavior works for as long as the feature gate is off.

Reply:
My understanding was that this question is specifically about having tests where you start with a cluster where the feature is disabled, and then roll out an enablement of the feature across the cluster (or vice versa) and so it is only important in the case where rolling out the feature has different failure modes from the case where the cluster is first started up with the feature already enabled.

Reply:
I usually interpret this section as "you have to have unit tests for code which depends on a feature gate with test cases for feature on resp. off".

Reply:
What Patrick says above, in general you want to make sure old behavior with the feature off, and new behavior with feature on. Usually, units or integration, since that's where it's the easiest to tweak the feature gates.